### PR TITLE
[RFC] Swift 4, simplifying response parsing, and no more Argo?

### PR DESCRIPTION
**Pull request by sharplet**
This is an experiment to see what it might be like if Swish drops support for Argo by requiring Swift 4 and making use of the new `JSONDecoder`. It also makes some changes to the way URL responses are handled, removing some optionality by encoding some assumptions described in the `URLSession.dataTask(with:completionHandler:)` documentation. See commits for more details.

### CODE REVIEW

**Comment by sharplet** [comment]
I'm going to make a `swift-4` branch and target this PR there, so we can keep breaking changes isolated from `master` for now.

File: Source/Extensions/Deprecations+Removals.swift [hunk 1]
```diff
@@ -1,24 +0,0 @@
...
-    fatalError("unavailable")
-  }
-}
-
-extension Result {
-  @available(*, unavailable, renamed: "Result.init(_:)")
-  public static func fromDecoded(_: Decoded<T>) -> Result<T, SwishError> {
-    fatalError("unavailable")
-  }
-}
```
Comment (gfontenot) [inline Source/Extensions/Deprecations+Removals.swift:24]: What do you think about leaving this file in place for future deprecations, but keeping it empty?

**Review by gfontenot — APPROVED** [review]
Sad to see Argo go here, but all in all this makes a lot of sense.

File: Source/Extensions/Deprecations+Removals.swift (hunk 1 above)
Comment (sharplet) [reply Source/Extensions/Deprecations+Removals.swift:24]: Sure

**Comment by njdehoog** [comment]
This looks good. Are there plans to make this part of a new release? It looks like it's been on hold for a while.

**Comment by klaaspieter** [comment]
@sharplet I rebased this for you. Okay to merge?

**Comment by sharplet** [comment]
@klaaspieter Thanks! I've fixed up the CI issues as well, so I'm going to go ahead and merge this.

**Comment by sharplet** [comment]
I'll look into putting out an official release later this week. In the meantime, this is available on `master`.

**Pull request closed without merging**